### PR TITLE
Watch pod events per cluster using non-blocking IO

### PR DESCRIPTION
**Pull request by msufa**
- monkey patch `kubeclient` to use a non-blocking http lib in watch streams
 - start a watcher per cluster on init
 - create/restart watchers when k8s clusters added/edited
 - remove obsolete classes from model

/cc @jonmoter @henders @sbrnunes 

### References
 - Jira link: 

### Risks
 - None

### CODE REVIEW

File: plugins/kubernetes/config/initializers/watchers.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
+require 'celluloid/current'
+require 'celluloid/io'
+require 'kubeclient'
+require 'http'
+require 'json'
+require 'logger'
+
+# monkey patching kubeclient to use non-blocking IO in watch streams
```
Comment (msufa) [inline plugins/kubernetes/config/initializers/watchers.rb]: This is the original for reference: https://github.com/abonas/kubeclient/blob/master/lib/kubeclient/watch_stream.rb

File: plugins/kubernetes/app/models/kubernetes/util.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require 'celluloid/current'
```
Comment (henders) [inline plugins/kubernetes/app/models/kubernetes/util.rb]: This is now in the lib/samson_plugin.rb file, but it should be harmless.

File: plugins/kubernetes/app/models/kubernetes/util.rb [hunk 3]
```diff
@@ -20,5 +22,20 @@ def self.log(message, extra_info = {})
...
+    def self.start_watcher(cluster)
+      watcher_name = pod_watcher_symbol(cluster)
+      Watchers::ClusterPodWatcher.supervise as: watcher_name, args: [cluster.client]
+    end
+
+    def self.restart_watcher(cluster)
+      watcher = Celluloid::Actor[pod_watcher_symbol(cluster)]
+      watcher.terminate if watcher and watcher.alive?
+      start_watcher(cluster)
+    end
```
Comment (henders) [inline plugins/kubernetes/app/models/kubernetes/util.rb]: This seems an odd place to put this logic.  Why not a class method in the ```Watchers::ClusterPodWatcher``` itself?

**Comment by henders** [comment]
You forgot to remove the line which causes the exception during deploy:
https://github.com/zendesk/samson/pull/645/files#diff-e25235b8653153a4ecb45b1cd3453a0dL22

:)

File: plugins/kubernetes/app/models/watchers/cluster_pod_watcher.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      else
+        false
+      end
+    end
+
+    def handle_notice(notice)
+      debug notice.to_s
+      return if handle_error(notice)
+      rc_name = notice.object.metadata.labels['replication_controller']
+      publish rc_name, notice if rc_name
```
Comment (henders) [inline plugins/kubernetes/app/models/watchers/cluster_pod_watcher.rb]: Should use parenthesis to make this a bit more readable ```publish(rc_name, notice) if rc_name```

File: plugins/kubernetes/config/initializers/watchers.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        end
+      end
+    end
+  end
+end
+
+Celluloid.logger = Rails.logger
+$CELLULOID_DEBUG = true
+
+Kubernetes::Cluster.all.each { |cluster| Kubernetes::Util::start_watcher(cluster) }
```
Comment (henders) [inline plugins/kubernetes/config/initializers/watchers.rb]: Only run this during web server execution:
https://github.com/zendesk/samson/blob/master/config/initializers/zz.rb#L1

File: plugins/kubernetes/config/initializers/watchers.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        @finished = true
+        @http.close unless @http.nil?
+      end
+
+      private
+
+      def get_client
+        !@http_options[:use_ssl] && @http_options[:basic_auth_user] && @http_options[:basic_auth_password] ?
+            HTTP.basic_auth(user: @http_options[:basic_auth_user], pass: @http_options[:basic_auth_password]) :
+            HTTP::Client.new
```
Comment (henders) [inline plugins/kubernetes/config/initializers/watchers.rb]: Use an if/else format here, multiline ternaries are a bit hard to parse

**Pull request merged**